resteasy-client-jetty/src/main/java/org/jboss/resteasy/client/jaxrs/engines/jetty/JettyClientEngine.java

```diff
@@ -112,112 +97,75 @@ public <T> CompletableFuture<T> submit(ClientInvocation request, boolean buffere
         return doSubmit(request, buffered, null, extractor);
     }
 
-    private <T> CompletableFuture<T> doSubmit(ClientInvocation invocation, boolean bufIn, InvocationCallback<T> callback,
+    private <T> CompletableFuture<T> doSubmit(ClientInvocation invocation, boolean buffered, InvocationCallback<T> callback,
             ResultExtractor<T> extractor) {
-        final boolean buffered;
-        if (!bufIn && extractor != null) {
-            if (!WARN_BUF.getAndSet(true)) {
-                Logger LOG = Logger.getLogger(JettyClientEngine.class);
-                LOG.error("TODO: ResultExtractor is synchronous and may not be used without buffering - forcing buffer mode.");
-            }
-            buffered = true;
-        } else {
-            buffered = bufIn;
-        }
+        final ExecutorService asyncExecutor = invocation.asyncInvocationExecutor();
 
         final Request request = client.newRequest(invocation.getUri());
         final CompletableFuture<T> future = new RequestFuture<T>(request);
-        // readEntity calls releaseConnection which calls cancel, so don't let that interrupt us
-        final AtomicBoolean completing = new AtomicBoolean();
 
         invocation.getMutableProperties().forEach(request::attribute);
         request.method(invocation.getMethod());
-        invocation.getHeaders().asMap().forEach((h, vs) -> vs.forEach(v -> request.header(h, v)));
+        request.headers(mutableHeaders -> invocation.getHeaders().asMap()
+                .forEach((h, vs) -> vs.forEach(v -> mutableHeaders.add(h, v))));
         configureTimeout(request);
         if (request.getAttributes().get(FOLLOW_REDIRECTS) == Boolean.FALSE) {
             request.followRedirects(false);
         }
 
         if (invocation.getEntity() != null) {
-            final ByteArrayOutputStream os = new ByteArrayOutputStream();
-            try {
-                invocation.writeRequestBody(os);
-            } catch (IOException e) {
-                future.completeExceptionally(e);
-                if (callback != null) {
-                    callback.failed(e);
+            final OutputStreamRequestContent contentOut = new OutputStreamRequestContent(
+                    invocation.getHeaders().getMediaType().toString());
```

Comment:
The `getMediaType().toString()` has potential to throw an NPE. We should check the `MediaType` is not `null` before invoking `.toString()`.